pkg/apis/core/validation/namespacedcloudprofile.go

```diff
@@ -7,17 +7,35 @@ package validation
 import (
 	apivalidation "k8s.io/apimachinery/pkg/api/validation"
 	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/utils/ptr"
 
 	"github.com/gardener/gardener/pkg/apis/core"
 	"github.com/gardener/gardener/pkg/utils"
 )
 
 // ValidateNamespacedCloudProfile validates a NamespacedCloudProfile object.
-func ValidateNamespacedCloudProfile(cloudProfile *core.NamespacedCloudProfile) field.ErrorList {
+func ValidateNamespacedCloudProfile(namespacedCloudProfile *core.NamespacedCloudProfile) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	allErrs = append(allErrs, apivalidation.ValidateObjectMeta(&cloudProfile.ObjectMeta, true, ValidateName, field.NewPath("metadata"))...)
-	allErrs = append(allErrs, validateNamespacedCloudProfileSpec(&cloudProfile.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, apivalidation.ValidateObjectMeta(&namespacedCloudProfile.ObjectMeta, true, ValidateName, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, validateParent(namespacedCloudProfile.Spec.Parent, field.NewPath("spec.parent"))...)
+
+	if namespacedCloudProfile.Spec.Kubernetes != nil {
+		allErrs = append(allErrs, validateOnlyExpirationsKubernetesVersions(namespacedCloudProfile.Spec.Kubernetes.Versions, field.NewPath("spec.kubernetes"))...)
+	}
+	allErrs = append(allErrs, validateOnlyExpirationsMachineImages(namespacedCloudProfile.Spec.MachineImages, field.NewPath("spec.machineImages"))...)
+
+	if len(namespacedCloudProfile.Spec.MachineTypes) > 0 {
+		allErrs = append(allErrs, validateMachineTypes(namespacedCloudProfile.Spec.MachineTypes, field.NewPath("spec.machineTypes"))...)
+	}
+	allErrs = append(allErrs, validateVolumeTypes(namespacedCloudProfile.Spec.VolumeTypes, field.NewPath("spec.volumeTypes"))...)
```

Comment:
Can you make this consistent, i.e. handle the `nil` and `len() == 0` cases in the validate functions.

Reply:
the problem is that e.g. for `validateMachineTypes`, as I am using the function of the `CloudProfile` validation, the restriction differs: In a `CloudProfile`, specifying a machine type is mandatory, whereas for a `NamespacedCloudProfile` it is not mandatory to provide a machine type.

Reply:
Understood. Does it make sense to separate/extract this a bit further, so that we have common validation logic for CP and NCP (in `utils.go`) and special constraints (e.g. machine type is mandatory)?

Reply:
extracted as common `ValidateMachineTypes` to `utils.go` and added `validateMandatoryMachineTypes` for use in `cloudprofile.go`.